cpp/models/abm/location.cpp

```diff
@@ -34,210 +33,166 @@ namespace abm
+uint32_t Location::get_number_persons()
 {
-    return (int)m_subpopulations.get_last_value()[size_t(s)];
+    return std::accumulate(m_cells.begin(), m_cells.end(), 0, [](uint32_t sum, auto cell) {
+        return sum + cell.m_persons.size();
+    });
 }
 
 /*
-For every location we have a transmission factor that is nomalized to m_capacity.volume / m_capacity.persons of 
-the location "Home", which is 66. We multiply this rate with the factor m_capacity.persons / m_capacity.volume of the 
-considered location. For the dependency of the transmission risk on the occupancy we use a linear approximation which 
-leads to: 66 * (m_capacity.persons / m_capacity.volume) * (m_num_persons / m_capacity.persons).
+For every cell in a location we have a transmission factor that is nomalized to m_capacity.volume / m_capacity.persons of 
+the location "Home", which is 66. We multiply this rate with the individual size of each cell to obtain a "space per person" factor.
 */
-ScalarType Location::compute_relative_transmission_risk()
+ScalarType Cell::compute_space_per_person_relative()
 {
-    if (m_capacity.volume != 0 && m_capacity_adapted_transmission_risk) {
-        return 66.0 * m_num_persons / m_capacity.volume;
+    if (m_capacity.volume != 0) {
+        return 66.0 / m_capacity.volume;
     }
     else {
         return 1.0;
     }
 }
 
-void Location::add_subpopulations_timepoint(const TimePoint& t)
+uint32_t Cell::get_subpopulation(TimePoint t, InfectionState state) const
+{
+    return count_if(m_persons.begin(), m_persons.end(), [&](observer_ptr<Person> p) {
+        return p->get_infection_state(t) == state;
+    });
+}
+
+uint32_t Location::get_subpopulation(TimePoint t, InfectionState state) const
+{
+    std::vector<Person*> loc_persons{};
+    for (auto&& cell : m_cells) {
+        for (auto p : cell.m_persons) {
+            loc_persons.push_back(p.get());
+        }
+    }
+    std::sort(loc_persons.begin(), loc_persons.end());
+    loc_persons.erase(std::unique(loc_persons.begin(), loc_persons.end()), loc_persons.end());
+    return count_if(loc_persons.begin(), loc_persons.end(), [&](Person* p) {
+        return p->get_infection_state(t) == state;
+    });
+}
+
+void Location::store_subpopulations(const TimePoint t)
 {
-    // Get the previous time point index.
-    // Since index starts from 0, we need to -1 from the current time point.
-    auto last_idx = m_subpopulations.get_num_time_points() - 1;
     m_subpopulations.add_time_point(t.days());
-    m_subpopulations.get_last_value() = m_subpopulations.get_value(last_idx);
+    std::array<int, size_t(InfectionState::Count)> subpopulations;
+    for (uint32_t i = 0; i < subpopulations.size(); ++i) {
+        subpopulations[i] = get_subpopulation(t, static_cast<InfectionState>(i));
+    }
+    m_subpopulations.get_last_value() =
+        Eigen::Map<const Eigen::VectorXi>(subpopulations.data(), subpopulations.size()).cast<ScalarType>();
 }
```

Comment:
you probably could build something with std::accumulate, where the "sum" is a vector. but I'm not sure it's worth it and if it's effficient.

Reply:
Should be faster and easier now since the location also saves the persons. Not sure if the performance is bad due to the duplication in adding and removing.